omit empty coupon fields

bobholt opened the pull request:
This makes optional arguments to CreateCoupon optional.

When these were not `omitempty`, we received `the provided xml was invalid` errors from Recurly's API. Additionally, the boolean arguments defaulted to `false` when not explicitly set to `true`, which was a surprise as Recurly assumes some of them default to `true` (e.g. `applies_to_all_plans`).

After this change, the XML is accepted by Recurly with expected defaults.

bobholt commented:
I don't have deep context into how Recurly treats previews vs. full subscriptions, so some questions:

- Do they document separate object types for real subscriptions and subscriptionPreviews, or is this to allow us to distinguish them?
- The `Subscription` and `SubscriptionPreview` objects look identical except the full `Subscription` doesn't have Tax-related fields. Is that accurate?
- Is this a change to the recurly lib that we anticipate blacklightcms would need to make, too? If so, can we upstream the change (contribute back to the library)?

codylee-splice commented:
> I don't have deep context into how Recurly treats previews vs. full subscriptions, so some questions:
> 
> * Do they document separate object types for real subscriptions and subscriptionPreviews, or is this to allow us to distinguish them?
> * The `Subscription` and `SubscriptionPreview` objects look identical except the full `Subscription` doesn't have Tax-related fields. Is that accurate?

There are separate object types but they're both both called `subscription`

Relevent Recurly APIs we're looking at:
https://dev.recurly.com/v2.22/docs/create-subscription
https://dev.recurly.com/v2.22/docs/preview-subscription

The major difference conceptually is that the *created* subscription has references to an invoice, address, etc. whereas the *preview* subscription has proposed values.

> * Is this a change to the recurly lib that we anticipate blacklightcms would need to make, too? If so, can we upstream the change (contribute back to the library)?

Maybe, probably? _I_ lack a good idea of how much we've changed this library so I'm not sure if we're doing it right. Also, the official Recurly Go library is out for Recurly v3 API so hopefully we won't have to maintain this library much/any longer.

bobholt commented:
> There are separate object types but they're both both called `subscription`

I'm not sure they are. It looks to me like they are the same type, but several fields are optional. In the OpenAPI docs for the v3 API I only see a single Subscription-like entity: https://github.com/recurly/recurly-client-go/blob/master/openapi/api.yaml#L15343

If we need the distinction in our code, I'd be cautious about introducing it into the here, and would instead create the separation of types in `api`. While I'm tempted to consider our fork throwaway if we're moving to the official Go API, I think we'll find that if we make this change here, and set up our code to _expect_ different types out of the lib, then when we switch to the official library and they're the same type, we might run into trouble.

That all said, if you need to ship, ship.

> I lack a good idea of how much we've changed this library so I'm not sure if we're doing it right.

At one point we had made very few changes here, and what we had changed, we upstreamed to `blacklightcms`. I'd have to look at the diff to see if we fell off that. If you continue forward with this change, I wouldn't upstream it.

codylee-splice commented:
> > There are separate object types but they're both both called `subscription`
> 
> I'm not sure they are. It looks to me like they are the same type, but several fields are optional.

Ah, that makes much more sense. I'll diff against blacklightcms and try to clean it up, accounting for variations.

> when we switch to the official library and they're the same type, we might run into trouble.

I don't think that'll be an issue because I don't see `subscriptions/preview` in v3 anyway, so we have to rewrite our code to use purchase preview in the switch, so create subscription and preview purchase return different object types in v3.

> That all said, if you need to ship, ship.
> 
> At one point we had made very few changes here, and what we had changed, we upstreamed to `blacklightcms`. I'd have to look at the diff to see if we fell off that.

I'll spend some time this morning seeing if I can unify them to good effect with the intent of minimal changes and upstreaming. If I don't see it as a viable solution I'll pull the trigger as is and make the claim that this fork should be EOLifed.

> If you continue forward with this change, I wouldn't upstream it.

Agreed.
